src/librustc/traits/select.rs

```diff
@@ -1143,15 +1148,15 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         let tcx = self.tcx();
         if self.can_use_global_caches(param_env) {
             let cache = tcx.evaluation_cache.hashmap.borrow();
-            if let Some(cached) = cache.get(&trait_ref) {
+            if let Some(cached) = cache.get(&param_env.and(trait_ref)) {
```

Comment:
ok, so, in this case the only purpose is to capture the reveal mode, correct?

Reply:
Yes, I can change to be literally that if e.g. it's a perf problem.

Reply:
I wouldn't expect that to have a perf impact, but might be worth including in a comment somewhere.